src/server/scripts/Northrend/Naxxramas/boss_sapphiron.cpp

```diff
@@ -359,56 +381,252 @@ class boss_sapphiron : public CreatureScript
+        {
+            if (_block)
+                return;
+            if (GetTarget()->isMoving())
+                return;
+            float x, y, z;
+            GetTarget()->GetPosition(x, y, z);
+            if (GameObject* block = GetTarget()->SummonGameObject(GO_ICEBLOCK, x, y, z, 0, 0, 0, 0, 0, 25))
+                _block = block->GetGUID();
+        }
+
+        void Register() override
+        {
+            AfterEffectApply += AuraEffectApplyFn(spell_sapphiron_icebolt_AuraScript::HandleApply, EFFECT_0, SPELL_AURA_MOD_STUN, AURA_EFFECT_HANDLE_REAL);
+            AfterEffectRemove += AuraEffectRemoveFn(spell_sapphiron_icebolt_AuraScript::HandleRemove, EFFECT_0, SPELL_AURA_MOD_STUN, AURA_EFFECT_HANDLE_REAL);
+            OnEffectPeriodic += AuraEffectPeriodicFn(spell_sapphiron_icebolt_AuraScript::HandlePeriodic, EFFECT_2, SPELL_AURA_PERIODIC_TRIGGER_SPELL);
+        }
+
+        ObjectGuid _block;
+    };
+
+    AuraScript* GetAuraScript() const override
+    {
+        return new spell_sapphiron_icebolt_AuraScript();
+    }
+};
+
+class spell_sapphiron_frost_breath : public SpellScriptLoader
+{
+    public:
+        spell_sapphiron_frost_breath() : SpellScriptLoader("spell_sapphiron_frost_breath") { }
+
+        class spell_sapphiron_frost_breath_SpellScript : public SpellScript
+        {
+            PrepareSpellScript(spell_sapphiron_frost_breath_SpellScript);
+
+            bool Validate(SpellInfo const* /*spell*/) override
+            {
+                return !!sSpellMgr->GetSpellInfo(SPELL_FROST_EXPLOSION);
+            }
+
+            void HandleTargets(std::list<WorldObject*>& targetList)
+            {
+                std::list<GameObject*> blocks;
+                if(GetCaster())
+                    GetCaster()->GetGameObjectListWithEntryInGrid(blocks, GO_ICEBLOCK, 200.0f);
+
+                std::vector<Unit*> toRemove;
+                toRemove.reserve(3);
+                std::list<WorldObject*>::iterator it = targetList.begin();
+                while (it != targetList.end())
+                {
+                    Unit* target = (*it)->ToUnit();
+                    if (!target)
+                    {
+                        it = targetList.erase(it);
                         continue;
+                    }
 
                     if (target->HasAura(SPELL_ICEBOLT))
                     {
-                        target->ApplySpellImmune(0, IMMUNITY_ID, SPELL_FROST_EXPLOSION, true);
-                        targets.push_back(target);
+                        it = targetList.erase(it);
+                        toRemove.push_back(target);
+                        continue;
+                    }
+                    
+                    bool found = false;
+                    for (GameObject* block : blocks)
+                        if (block->IsInBetween(GetCaster(), target, 2.0f) && GetCaster()->GetExactDist2d(block) + 5 >= GetCaster()->GetExactDist2d(target))
```

Comment:
This is a good start. However, this isnt blizzlike. This spell should work by LOS in reference to Saphiron. Also the spell SPELL_FROST_BREATH is linked to SPELL_FROST_EXPLOSION. Both are cast at the same time. Dont ask me why however, I dont know :/